stdlib/asm/crypto/dsa/rpo_falcon512.masm

```diff
@@ -218,182 +207,83 @@ export.load_h_s2_and_product.4
         push.M u32lt assert
         push.M u32lt assert
 
+        horner_eval_base
+
         hperm
     end
 
-    # 3) Load saved claimed hash of h and compare
+    # 4) Load PK, the saved claimed hash of h, and compare
     loc_loadw.0
     movup.4 assert_eq
     movup.3 assert_eq
     movup.2 assert_eq
     assert_eq
+    # => [C, ptr, alpha_inv, h(tau)_1, h(tau)_0, ...]
 
-    # 4) Load s2 (Due to the final norm test we do not need to range check the s2 coefficients)
-    padw padw
-    repeat.64
-        adv_pipe hperm
-    end
-
-    # 5) Load claimed h * s2 in Z_Q[x]
-    repeat.128
-        adv_pipe hperm
-    end
-
-    # 6) Return the challenge point and the incremented pointer
-    exec.rpo::squeeze_digest
-    drop drop
-    #=> [tau1, tau0, ptr + 512*4]
-end
-
-#! Checks that pi == h * s2 in Z_Q[x] by evaluating both sides at a random point.
-#! The procedure takes as input a pointer h_ptr to h. The other two polynomials
-#! are located at h_ptr + 512, for s2, and h_ptr + 1024, for pi. The procedure takes
-#! also a pointer zeros_ptr to a region of memory [zeros_ptr, zeros_ptr + 512*4)
-#! and a pointer tau_ptr to powers of the random point we are evaluating at stored
-#! as [a_i, b_i, x, x] where (a_i, b_i) := tau^i for i in [0, 1023].
-#! The procedure returns () if the check passes, otherwise it raises an exception
-#! related to an unsatisfied assertion.
-#!
-#! Input: [h_ptr, zeros_ptr, tau_ptr, ...]
-#! Output: [...]
-#!
-#! Cycles: 2504
-export.probabilistic_product.16
-    # 1) Save the pointers
-    push.0 movdn.3
-    loc_storew.0
-
-    # 2) Compute the evaluation of the h polynomial at the random challenge
-
-    # Accumulator to compute h(tau)
-    padw
-
-    # For mem_stream
+    # 5) Load s2 and evaluate at tau_inv (Due to the final norm test we do not need to range check the s2 coefficients)
+    
+    ## a) Reset the accumulator, update the pointers and set up the state of the hasher
+    push.0 movdn.6
+    push.0 movdn.6
     padw padw
 
-    # Compute h(tau)
+    ## b) Load and evaluate s2 at tau_inv
```

Comment:
Here too we don't need to range-check the coefficients, right?

Reply:
Once `h` and `s2` are range checked, we know for sure that the coefficients of the product polynomial cannot overflow the size of the Miden VM modulus, and this is all we need to impose on the product polynomial coefficients.